### ISSUE TITLE
Remove "Lazy" ExtendedPdo in 3.x

### ISSUE THREAD

**Issue by pmjones** [auraphp/Aura.Sql#119]
It occurs to me that there are better ways to manage laziness of connections, than by having the connection manage itself. Providing a Factory or Container that only creates an ExtendedPdo on demand, rather than the LazyExtendedPdo instance creating its own internal connection on demand, would support a better separation of concerns. While this reduces the ability of older codebases to migrate directly to 3.x to get the benefits of lazy connection, the 2.x lazy connection remains available for that purpose; further, it streamlines and simplifies the ExtendedPdo logic itself.

@pavarnos @harikt and others: thoughts?

EDIT: Proof of concept branch in https://github.com/auraphp/Aura.Sql/tree/nonlazy .

**Comment by harikt** [comment]
I am in the late nights, so I will keep this for a fresh day to think :) .

Thank you.

**Comment by mbrevda** [comment]
This can cause issues when unit testing. I.e. if my class has an Sql instance passed in on construction, which tries to connect on jnstanciation, my test would fail if there is no DB available. But what if my test doesn't require a DB at All?

**Comment by pmjones** [comment]
@mbrevda Fair point. For testing purposes, if a DB is not needed and not actually available, might it be worthwhile to pass a "FakeExtendedPdo" that overrides the constructor to not-connect at all?

**Comment by mbrevda** [comment]
I think that is an oversimplification that went past the red line and is no back on over complicated territory. 

The test might be to confirm that the ExtendedPdo class is actually available - and yet we still require no connection.

**Comment by pmjones** [comment]
(/me nods)

> The test might be to confirm that the ExtendedPdo class is actually available - and yet we still require no connection.

I presumed, but did not explicitly say, that FakeExtendedPdo would extend ExtendedPdo to override the constructor. Confirming that `$fakeExtendedPdo instanceof ExtendedPdo` should still pass in that case.

Keep the Devil's advocacy coming.

**Comment by pavarnos** [comment]
what about making LazyExtendedPdo accept a callable in the constructor instead of dsn, username, password etc? This allows 
- laziness without having to use the connection locator (which is good for smaller projects), 
- allows decoration of an existing PDO (to cover that use-case which I completely overlooked in #116)
- simplifies LazyExtendedPdo: fewer private member variables
- would allow a mock PDO object to be injected for unit tests (where the mock is generated as needed by Mockery, phpUnit etc)

eg LazyPdoConnection:

``` php
public function __constructor(callable $getPdoConnection) {
    $this->getPdoConnection = $getPdoConnection;
}

public function connect() {
    if (empty($this->pdo)) {
        $this->pdo = call_user_func($this->getPdoConnection);
    }
}
```

for unit tests eg in phpUnit

``` php
    $pdo = new LazyPdoConnection( function () { return $this->getMock('\Aura\Sql\ExtendedPdo');} ));
```

**Comment by pavarnos** [comment]
ps. the variadics in query() are much cleaner. very nice! good to get rid of that ugly code... i'd been focussing on php5.5 and earlier compatibility. Good to be free of that...

**Comment by pmjones** [comment]
> which is good for smaller projects

(/me nods)

I am not sold on the idea of removing laziness from ExtendedPdo, but I do find it appealing, esp. given my development experiences of the past few years. Even in smaller projects, there has often (even usually) been a DI container of some sort that provides laziness. Objects needing a PDO connection have that connection created only as the object itself gets instantiated.

**Comment by pavarnos** [comment]
(after thinking about it for a day or so) imho in a generic toolkit like this, it "might be handy" to have 
- a decorating / proxying class as a leg-up for users so they don't have to roll their own. A wee bit of documentation could show how it can be repurposed into a mock, a lazy loader, an existing-pdo-decorator (for when you need compatibility with another framework) 
- a NullExtendedPdo or FakeExtendedPdo where every function returns '', 0, [] as appropriate for unit tests (as suggested by @mbrevda)

its a toss up between simplicity / minimalism and broader usefulness. i could be convinced either way...

**Comment by harikt** [comment]
If my understanding is correct phpunit out of the box now support https://github.com/phpspec/prophecy . So you can actually mock things easily for testing.

Regarding removing of laziness, I am probably not understanding what is the reason we need to remove the laziness.

**Comment by harikt** [comment]
can we do in the `connect()` method calling the `parent::__construct()` if not called ? So we don't call the `new ExtendedPdo` in `LazyExtendedPdo` ?

**Comment by pavarnos** [comment]
@harikt yes if we pass in a callable instead of the dsn, user, password etc. See my earlier comment above

**Comment by harikt** [comment]
@pavarnos I think you are talking about [this comment](https://github.com/auraphp/Aura.Sql/issues/119#issuecomment-195800970) . But that is a bit different with what I mentioned. 

To clear up what I am telling is  

``` php
class ExtendedPdo extends PDO ...
{
    protected $connected = false;

    // override pdo construct
    public function __construct()

    public function connect()
    {
        if (! $this->connected) {
            parent::__construct(call parent construct of PDO here );
            $this->connected = true;
        }
    }
}
```

So we are lazy with the `ExtendedPdo` itself. This is what I tried : ~~https://3v4l.org/hchAI~~ . 

Updating with https://3v4l.org/R3RQ5

**Comment by harikt** [comment]
UPDATE : ~~I see one problem, we cannot make use of existing pdo connection in that case. Eg : https://github.com/auraphp/Aura.Sql/pull/32~~ assumption was wrong.

**Comment by pavarnos** [comment]
Hopefully #122 answers some of these issues? using `setPdoFactory` lets you inject an existing pdo, lave a lazy profiled extended pdo and other useful combinations.

**Comment by pmjones** [comment]
@harikt That's very clever, running the constructor from within connect(). I see @pavarnos mentioned it elsewhere too. Normally I don't like "clever" but it seems like it might work well.

**Comment by pmjones** [comment]
@pavarnos with @harikt's "connect/construct" approach in mind, perhaps an added "DecoratedPdo" becomes the thing, where an existing PDO connection is passed to the constructor.

**Comment by pmjones** [comment]
@pavarnos I'll see your #122 and raise you the commits in the `nonlazy` branch starting earlier today: https://github.com/auraphp/Aura.Sql/commits/nonlazy

**Comment by pavarnos** [comment]
Yes. Was having trouble visualising how it would work

**Comment by pmjones** [comment]
So, the problem with the connect/construct approach, in combination with a DecoratedPdo, is that at some point you have to proxy everything to the underlying PDO object being decorated. Scrutinizer does not like that so much; ExtendedPdo gets an "A" in nonlazy, but the proxied version in the "relazy" branch gets a "C". (The relazy ExtendedPdo decorates an underlying PDO object.)

https://scrutinizer-ci.com/g/auraphp/Aura.Sql/?branch=nonlazy

https://scrutinizer-ci.com/g/auraphp/Aura.Sql/?branch=relazy

Is that a deal-breaker for anyone?

(One note about those two branches: they use a refactored version of @pavarnos profiler/logger work that extracts the Profiler to its own object, obviating the need for a separate ProfiledExtendedPdo.)

**Comment by pavarnos** [comment]
I like the new Profiler class (and interface). Makes it easy to switch in a new one if you don't like how the bundled one works.

Not really liking the profiler as part of ExtendedPdo because I don't want to pay the overhead of calling a profiler for each function when most of the time I will not need it (especially in production). 
eg

``` php
 public function beginTransaction()
    {
        $this->profiler->start(__FUNCTION__);
        $result = parent::beginTransaction();
        $this->profiler->finish();
        return $result;
    }
```

if you remove the profiler calls you don't need this function at all: you can just rely on the parent method in PDO. 

How well do you know the primary users of this component? Do most of them have the profiler switched on all the time? My vote is for the "naked" ExtendedPdo and a separate ProfiledExtendedPdo so that your DI container can inject it when you need it, and you don't have to pay for it when you don't need it.

re decorating: I feel no strong pull either way as long as the DecoratedPdo is separate from ExtendedPdo. Handy to have a decorator class in a component like this: adds flexibility. Metrics from scrutinizer are just a guide: its OK to ignore them for a "greater purpose" (imho).

tl;dr: my vote is to make ExtendedPdo as simple and lean as possible because its what most people will use most of the time.

thanks for asking btw...

**Comment by pmjones** [comment]
>  I don't want to pay the overhead of calling a profiler for each function when most of the time I will not need it (especially in production)

Totally understood; it _looks_ like a lot. However:  I have not benchmarked "this exact implementation" but I opine from "past related benchmarking" that the overhead of the extra calls to the profiler will be a rounding error compared to the time it takes to run the related query -- or even to the time it takes to just make the connection in the first place.  You get what I mean there?

> Metrics from scrutinizer are just a guide: its OK to ignore them for a "greater purpose" (imho).

Fair. And it's a hell of a lot better than the 6.29 the 2.x branch has. :-)

**Comment by harikt** [comment]
I also agree with @pavarnos thoughts here. And I do know @pmjones have lots of experience with the benchmarks and it may be very few micro / nano seconds for the function to execute. 

But still, probably it may be good to keep the profiler separate though.
I do understand for quick testing if profiler is within, it may be easy. So now I am confused :-) .

**Comment by pmjones** [comment]
All right, a couple of "executive decisions":
- **Not** removing laziness after all. A couple of experiments showed that it could be retained, with good marks from Scrutinizer.
- **Not** using a separate ProfiledExtendedPdo, at least not at this time.

The work is in the "relazy" branch, which I'll merge to 3.x shortly.

Thanks, everyone, for your patience with my working-things-out, and for your valuable critiques.
